### PR TITLE
Delegate method to restrict movement

### LINKED ISSUE #2457: Let MGLMapViewDelegate cancel region changes
`-[MGLMapViewDelegate mapView:regionWillChangeAnimated:]` should return a Boolean, so that delegates can cancel impending viewport changes by return `NO`. If the viewport change was initiated programmatically, perhaps the return value should be ignored, but it should at least be honored for viewport changes initiated by gesture recognizers.

/cc @friedbunny @incanus

### PR DESCRIPTION
**Pull request by 1ec5**
Added a way for the delegate to restrict where the user can move within the map using gestures. This approach allows the developer to restrict to a non-rectangular region or vary the region by zoom level or pitch as necessary, such as to prevent the user from seeing beyond a certain region regardless of the degree of tilt.

This PR intentionally keeps things simple to unblock some key use cases. Future improvements could include implementing a “rubber band” effect, allowing the delegate to “redirect” the camera to some other camera (https://github.com/mapbox/mapbox-gl-native/issues/2457#issuecomment-222051074), and automatically restricting the camera to downloaded regions (#5042).

Before we can land this feature:

* [ ] Restrict all the other gestures on iOS: https://github.com/mapbox/mapbox-gl-native/pull/5584#issuecomment-230885270
* [ ] Add a warning about performance: https://github.com/mapbox/mapbox-gl-native/pull/5584#r69783094
  * Or maybe find a way to call the method less frequently?
  * Or calculate the final camera beforehand and avoid the set-and-revert dance?

Fixes #2457.

/cc @boundsj @frederoni @friedbunny

### CODE REVIEW

**Comment by 1ec5** [comment]
I think it’s too late for this feature to make the impending v3.3.0 final release, but we could easily backport it to v3.3.1 if it takes awhile to release v3.4.0.

**Comment by 1ec5** [comment]
Note that this API does not affect programmatic movement of the viewport, such as via `-flyToCamera:…`, so we won’t run into mapbox/mapbox-gl-js#2521.

File: platform/macos/src/MGLMapViewDelegate.h [hunk 1]
```diff
@@ -61,6 +61,25 @@ NS_ASSUME_NONNULL_BEGIN
...
 - (void)mapView:(MGLMapView *)mapView cameraDidChangeAnimated:(BOOL)animated;
 
+/**
+ Asks the delegate whether the map view should be allowed to change from the
+ existing camera to the new camera in response to a user gesture.
+ 
+ This method is called as soon as the user gesture is recognized. It is not
+ called in response to a programmatic camera change, such as by setting the
+ `centerCoordinate` property or calling `-flyToCamera:completionHandler:`.
+
```
Comment (friedbunny) [inline platform/macos/src/MGLMapViewDelegate.h]: Should we add the standard, “this method is called many times during gesticulation so you should keep it as lightweight as possible”?

**Comment by friedbunny** [comment]
This only seems to prevent simple pans, so far. With the new delegate method set to always return `NO`, it’s still possible to move the map via flick, pinch + pan, and rotate + pan.

**Comment by 1ec5** [comment]
> With the new delegate method set to always return `NO`, it’s still possible to move the map via flick, pinch + pan, and rotate + pan. Quick zoom also works.

Oh, you’re right! I took care of all the gesture recognizers for macOS but forgot to follow up for iOS.

**Comment by 1ec5** [comment]
Another caveat about this approach is that, if the delegate decides to block the camera change, the camera actually does change (even sending the `-mapView:didChangeRegionAnimated:` message to the delegate) before the map view resets the camera.

**Comment by Zakay** [comment]
👍

**Comment by maciekish** [comment]
Any chance to see this in v3.3.1 please?

**Comment by 1ec5** [comment]
v3.3.1 has already been released. This work is currently targeted for v3.4.0. However, lately I’ve been focused on the runtime styling API ahead of the first beta, since it has the potential for a lot more API churn.

**Comment by 1ec5** [comment]
> This feels a bit more useful if we can determine more about the viewport than just the newCamera center coordinate

Yep, there’s definitely a lot more potential here. Future smarts probably belong in mbgl, whereas I tried to keep things light in this PR. For example, if the use case is to prevent a certain tile from being shown (with rotation and pitch gestures enabled), there needs to be some nearest-point-in-polygon logic in mbgl, but for now, developers can bring their own implementation of that logic.

**Comment by maciekish** [comment]
This does not work when flicking the map. Delegate is never called.

**Comment by 1ec5** [comment]
> This does not work when flicking the map. Delegate is never called.

Yes, this is on my to-do list, per https://github.com/mapbox/mapbox-gl-native/pull/5584#issuecomment-230885270. All the gesture recognizers are hooked up on macOS, in case you want to try it out in the meantime.

**Comment by Sealos** [comment]
Any chance for this feature in 3.4.0?

**Comment by boundsj** [comment]
Moved this to 3.4.1

**Comment by boundsj** [comment]
Moving to 3.5.0 cc @1ec5

**Comment by boundsj** [comment]
@incanus can you please have another look at this and release your hold on it if it looks ok?

File: platform/ios/CHANGELOG.md [hunk 2]
```diff
@@ -99,6 +99,12 @@ Mapbox welcomes participation and contributions from everyone. Please read [CONT
 * Fixed an issue that caused an assertion failure if a `MGLShapeCollection` (a GeoJSON GeometryCollection) was created with an empty array of shapes. ([#7632](https://github.com/mapbox/mapbox-gl-native/pull/7632))
 * Improved the precision of annotations at zoom levels greater than 18. ([#5517](https://github.com/mapbox/mapbox-gl-native/pull/5517))
 
+### User interactivity
+
+* Added a method to MGLMapViewDelegate, `-mapView:shouldChangeFromCamera:toCamera:`, that you can implement to restrict which parts the user can navigate to using gestures. ([#5584](https://github.com/mapbox/mapbox-gl-native/pull/5584))
```
Comment (boundsj) [inline platform/ios/CHANGELOG.md]: This changelog entry should be moved to reflect the actual release it is going out in. Alternatively, this PR could be retargeted to the `master` branch and this entry could go in the "master" section. cc @1ec5

File: platform/macos/src/MGLMapView.mm [hunk 3]
```diff
@@ -1372,15 +1382,25 @@ - (void)handlePanGesture:(NSPanGestureRecognizer *)gestureRecognizer {
...
                 _mbglMap->setBearing(newDirection, center);
+                didChangeCamera = YES;
                 [self didChangeValueForKey:@"direction"];
             }
             if (self.pitchEnabled) {
                 _mbglMap->setPitch(_pitchAtBeginningOfGesture + delta.y / 5, center);
+                didChangeCamera = YES;
+            }
+            
+            if (didChangeCamera
```
Comment (boundsj) [inline platform/macos/src/MGLMapView.mm]: @fabian-guerra @1ec5 is there a reason we don't move this up and short circuit the bearing and pitch changes to the map completely?

File: platform/ios/CHANGELOG.md (hunk 2 above)
Comment (1ec5) [inline platform/ios/CHANGELOG.md]: This PR is already targeted at master, so this blurb needs to be moved up to “Other changes” under the existing master section.

File: platform/macos/src/MGLMapView.mm (hunk 3 above)
Comment (fabian-guerra) [inline platform/macos/src/MGLMapView.mm]: I realized that we change `didChangeCamera` only when `pan.state == UIGestureRecognizerStateChanged` this introduced a bug that caused the camera to move when `pan.state == UIGestureRecognizerStateEnded` I think we can remove this variable and do the short circuit instead. What are your thoughts @1ec5?

Comment (1ec5) [inline platform/macos/src/MGLMapView.mm]: I alluded to this up above:

> Or calculate the final camera beforehand and avoid the set-and-revert dance?

The reason I originally implemented it this way was that I expected to eventually implement a rubber-banding effect, in which we would have to set the camera (with animation) then revert. But we can backfill that behavior later without affecting the API, so short-circuiting should be fine.

File: platform/ios/src/MGLMapView.mm [hunk 4]
```diff
@@ -1337,6 +1337,12 @@ - (void)handlePinchGesture:(UIPinchGestureRecognizer *)pinch
 
         [self unrotateIfNeededForGesture];
     }
+    
+    if ([self.delegate respondsToSelector:@selector(mapView:shouldChangeFromCamera:toCamera:)]
+        && ![self.delegate mapView:self shouldChangeFromCamera:oldCamera toCamera:self.camera])
+    {
+        self.camera = oldCamera;
+    }
```
Comment (incanus) [inline platform/ios/src/MGLMapView.mm]: This seems to happen too late in the process? i.e. `setScale()` has already been called above, versus the way the way other gestures are handled in this PR.

File: platform/macos/src/MGLMapView.mm (hunk 3 above)
Comment (1ec5) [inline platform/macos/src/MGLMapView.mm]: The other reason I used a set-and-revert strategy was simplicity: we don’t know `newCamera` until we perform the transformation, and `mbgl::Map` lacks a “dry run” version of these transformation methods.

File: platform/ios/src/MGLMapView.mm (hunk 4 above)
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: This is consistent with all the gesture handlers I implemented on macOS, as well as `-handlePanGesture:` above, but you’re right that it’s inconsistent with `-handleRotateGesture:` below.

File: platform/ios/src/MGLMapView.mm [hunk 5]
```diff
@@ -1364,10 +1380,16 @@ - (void)handleRotateGesture:(UIRotationGestureRecognizer *)rotate
             newDegrees = fminf(newDegrees,  30);
             newDegrees = fmaxf(newDegrees, -30);
         }
-
-        _mbglMap->setBearing(newDegrees, mbgl::ScreenCoordinate { centerPoint.x, centerPoint.y });
-
-        [self notifyMapChange:mbgl::MapChangeRegionIsChanging];
+        
+        if ([self.delegate respondsToSelector:@selector(mapView:shouldChangeFromCamera:toCamera:)]
+            && ![self.delegate mapView:self shouldChangeFromCamera:oldCamera toCamera:self.camera])
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: `newCamera` equals `oldCamera` because we haven’t called `mbgl::Map::setBearing()` yet. `newCamera` should instead reflect the camera _after_ applying the rotation.

File: platform/macos/CHANGELOG.md [hunk 6]
```diff
@@ -67,6 +67,7 @@ This version of the Mapbox macOS SDK corresponds to version 3.4.0 of the Mapbox
 * Improved style parsing performance. ([#6170](https://github.com/mapbox/mapbox-gl-native/pull/6170))
 * Improved feature querying performance. ([#6514](https://github.com/mapbox/mapbox-gl-native/pull/6514))
 * Fixed an issue where shapes that cannot currently be visually represented as annotations were still shown on the map as point annotations. ([#6764](https://github.com/mapbox/mapbox-gl-native/issues/6764))
+* Raster tiles such as those from Mapbox Satellite are now cached, eliminating flashing while panning back and forth. ([#7091](https://github.com/mapbox/mapbox-gl-native/pull/7091))
```
Comment (1ec5) [inline platform/macos/CHANGELOG.md]: Did you mean to add this entry?

File: platform/ios/src/MGLMapView.mm [hunk 7]
```diff
@@ -1615,6 +1656,12 @@ - (void)handleTwoFingerDragGesture:(UIPanGestureRecognizer *)twoFingerDrag
         [self notifyGestureDidEndWithDrift:NO];
         [self unrotateIfNeededForGesture];
     }
+    
+    if ([self.delegate respondsToSelector:@selector(mapView:shouldChangeFromCamera:toCamera:)]
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: Unless you check the gesture recognizer's state (or didChangeCamera), the camera will be reset to oldCamera before the gesture is complete.

File: platform/macos/CHANGELOG.md (hunk 6 above)
Comment (fabian-guerra) [inline platform/macos/CHANGELOG.md]: eh.. no

File: platform/ios/src/MGLMapView.mm [hunk 8]
```diff
@@ -1533,17 +1560,25 @@ - (void)handleTwoFingerTapGesture:(UITapGestureRecognizer *)twoFingerTap
...
-
-        mbgl::ScreenCoordinate center(gesturePoint.x, gesturePoint.y);
-        _mbglMap->scaleBy(0.5, center, MGLDurationInSecondsFromTimeInterval(MGLAnimationDuration));
-
-        __weak MGLMapView *weakSelf = self;
-
-        [self animateWithDelay:MGLAnimationDuration animations:^
+        MGLMapCamera *oldCamera = self.camera;
+        if ([self.delegate respondsToSelector:@selector(mapView:shouldChangeFromCamera:toCamera:)]
+            && ![self.delegate mapView:self shouldChangeFromCamera:oldCamera toCamera:self.camera])
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: self.camera at this point is equal to oldCamera. We need to pass in the destination camera as the third parameter. Could you create a copy of the camera and modify it to reflect the result after zooming out?

File: platform/ios/CHANGELOG.md [hunk 9]
```diff
@@ -99,6 +99,12 @@ Mapbox welcomes participation and contributions from everyone. Please read [CONT
...
+* The UITapGestureRecognizer on MGLMapView that is used for selecting annotations now fails if a tap does not select an annotation. ([#7246](https://github.com/mapbox/mapbox-gl-native/pull/7246))
```
Comment (1ec5) [inline platform/ios/CHANGELOG.md]: Can you undo all these unrelated changes?

File: platform/macos/CHANGELOG.md [hunk 10]
```diff
@@ -81,6 +82,10 @@ This version of the Mapbox macOS SDK corresponds to version 3.4.0 of the Mapbox
 * Fixed an issue that caused an assertion failure if a `MGLShapeCollection` (a GeoJSON GeometryCollection) was created with an empty array of shapes. ([#7632](https://github.com/mapbox/mapbox-gl-native/pull/7632))
 * Improved the precision of annotations at zoom levels greater than 18. ([#5517](https://github.com/mapbox/mapbox-gl-native/pull/5517))
 
+### User interactivity
+
+* The NSClickGestureRecognizer on MGLMapView that is used for selecting annotations now fails if a click does not select an annotation. ([#7246](https://github.com/mapbox/mapbox-gl-native/pull/7246))
```
Comment (1ec5) [inline platform/macos/CHANGELOG.md]: There are still a number of unrelated changes in both changelogs.

File: platform/ios/src/MGLMapView.mm [hunk 11]
```diff
@@ -1364,10 +1377,10 @@ - (void)handleRotateGesture:(UIRotationGestureRecognizer *)rotate
             newDegrees = fminf(newDegrees,  30);
             newDegrees = fmaxf(newDegrees, -30);
         }
-
+
```
Comment (incanus) [inline platform/ios/src/MGLMapView.mm]: Nitpick: get rid of extraneous whitespace here.

File: platform/ios/src/MGLMapView.mm [hunk 12]
```diff
@@ -1607,14 +1669,21 @@ - (void)handleTwoFingerDragGesture:(UIPanGestureRecognizer *)twoFingerDrag
         CGPoint centerPoint = [self anchorPointForGesture:twoFingerDrag];
 
         _mbglMap->setPitch(pitchNew, mbgl::ScreenCoordinate { centerPoint.x, centerPoint.y });
-
+        didCameraChanged = YES;
```
Comment (incanus) [inline platform/ios/src/MGLMapView.mm]: Nitpick: should probably call this `didChangeCamera`.

**Comment by 1ec5** [comment]
> Do we care that various metrics are still reported even when the old camera is reset, as the viewport would not end up changing but the change would be recorded?

Given the gesture lifecycle, this is an inevitable consequence of our set-and-revert strategy. I think it's OK, though, because the events are meant to track user intent, not necessarily tile views.

**Comment by 1ec5** [comment]
Indeed, the delegate method is still unable to stop drift-panning/zooming/rotating on iOS. Consider this scenario:

1. The delegate method is implemented so that the user may not pan beyond the antimeridian.
1. The user quickly pans westward toward the antimeridian, letting go before reaching it, but there's enough momentum that the map continues drift-panning.
1. Drift-pannning is implemented by calling mbgl::Map::panBy() with a duration. The animation runs asynchronously, so the delegate method gets called at the moment the animation begins, with a newCamera equal to oldCamera.
1. The delegate method returns YES because newCamera (which incorrectly matches oldCamera) still has a longitude greater than −180°.

The only gesture recognizer that currently handles this scenario correctly is the double-tap gesture recognizer, because it predicts the final camera instead of performing the set-and-revert strategy. In order to fix this issue comprehensively, we'd have to follow a similar approach with all the gesture recognizers. The code will be a little messier, but not by much.

The macOS implementation is unaffected by this issue, because the drift-panning animation is driven by AppKit instead of mbgl. The gesture recognizer method gets called incrementally after the user lets go.

File: platform/ios/src/MGLMapView.mm [hunk 13]
```diff
@@ -1508,18 +1573,36 @@ - (void)handleDoubleTapGesture:(UITapGestureRecognizer *)doubleTap
...
-
-        mbgl::ScreenCoordinate center(gesturePoint.x, gesturePoint.y);
-        _mbglMap->scaleBy(2, center, MGLDurationInSecondsFromTimeInterval(MGLAnimationDuration));
-
-        __weak MGLMapView *weakSelf = self;
-
-        [self animateWithDelay:MGLAnimationDuration animations:^
+        MGLMapCamera *oldCamera = self.camera;
+        
+        mbgl::EdgeInsets padding = MGLEdgeInsetsFromNSEdgeInsets(self.contentInset);
```
Comment (boundsj) [inline platform/ios/src/MGLMapView.mm]: nitpick: these several lines about getting padding, camera options, setting a zoom adjustment, clamping, and getting a `toCamera` from camera options are repeated in the `handleDoubleTapGesture:`, `handleTwoFingerTapGesture:` and `handleQuickZoomGesture:` implementations. This is similar to what we had before where the previous (less complex because no estimate) logic was repeated, but this PR could be a chance to refactor this into a private utility method. I think anything we can do to make the gesture handling more readable and easier to refactor in the future would be 👍 so that is why I mention this.

File: platform/ios/src/MGLMapView.mm [hunk 14]
```diff
@@ -1321,14 +1353,26 @@ - (void)handlePinchGesture:(UIPinchGestureRecognizer *)pinch
...
+        zoom = log2(newScale);
+        toCamera = [self currentCameraWithEstimatedZoom:zoom];
+        
+        if ([self.delegate respondsToSelector:@selector(mapView:shouldChangeFromCamera:toCamera:)]
+            && ![self.delegate mapView:self shouldChangeFromCamera:oldCamera toCamera:toCamera])
         {
-            _mbglMap->setScale(newScale, mbgl::ScreenCoordinate { centerPoint.x, centerPoint.y }, MGLDurationInSecondsFromTimeInterval(duration));
+            self.camera = oldCamera;
+            drift = NO;
+        }else {
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: Nit: missing space before `else`.

File: platform/ios/src/MGLMapView.mm [hunk 15]
```diff
@@ -1619,6 +1748,73 @@ - (void)handleTwoFingerDragGesture:(UIPanGestureRecognizer *)twoFingerDrag
...
+        width/2.,
+        height/2.,
+    };
+    
+    mbgl::ScreenCoordinate centerOffset = {
+        endPoint.x,
+        -endPoint.y,
+    };
+    
+    mbgl::ScreenCoordinate centerPoint = screenCenter - centerOffset;
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: You can simplify this logic by creating a `CGPoint` directly instead of a `ScreenCoordinate`:

```objc
CGPoint centerPoint = CGPointMake(CGRectGetMidX(self.bounds), CGRectGetMidY(self.bounds));
```

File: platform/ios/src/MGLMapView.mm [hunk 16]
```diff
@@ -1619,6 +1748,73 @@ - (void)handleTwoFingerDragGesture:(UIPanGestureRecognizer *)twoFingerDrag
         [self notifyGestureDidEndWithDrift:NO];
         [self unrotateIfNeededForGesture];
     }
+
+}
+
+- (MGLMapCamera *)currentCameraWithEstimatedEndPoint:(CGPoint)endPoint panGesture:(UIPanGestureRecognizer *)pan
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: `pan` is unused.

A better place for this logic would be a private `-[MGLMapCamera cameraByPanningWithTranslation:inMapView:]` (which would have parameters similar to `-[UIPanGestureRecognizer setTranslation:inView:]`). You’d need to create a new MGLMapCamera_Private.h for these methods.

This way you can also share this and similar methods between iOS and macOS. The downside is that our non-UI code would have some dependency on our UI code, but that won’t introduce any practical problems unless we decide to split the Darwin code into a separate framework down the line.

File: platform/ios/src/MGLMapView.mm [hunk 17]
```diff
@@ -1619,6 +1748,73 @@ - (void)handleTwoFingerDragGesture:(UIPanGestureRecognizer *)twoFingerDrag
...
+
+- (MGLMapCamera *)currentCameraWithEstimatedPitch:(CGFloat)pitch anchorPoint:(CGPoint)anchorPoint
+{
+    mbgl::EdgeInsets padding = MGLEdgeInsetsFromNSEdgeInsets(self.contentInset);
+    mbgl::CameraOptions currentCameraOptions = _mbglMap->getCameraOptions(padding);
+    
+    MGLMapCamera *camera;
+    
+    mbgl::ScreenCoordinate anchor = mbgl::ScreenCoordinate { anchorPoint.x, anchorPoint.y };
+    currentCameraOptions.pitch = pitch * mbgl::util::DEG2RAD;
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: Pitch has no anchor point – it’s always the center of the view, disregarding any padding – so this method only needs to copy the camera and set its pitch to the new value.

File: platform/ios/src/MGLMapView.mm [hunk 18]
```diff
@@ -1619,6 +1748,73 @@ - (void)handleTwoFingerDragGesture:(UIPanGestureRecognizer *)twoFingerDrag
...
+    
+    mbgl::ScreenCoordinate centerPoint = screenCenter - centerOffset;
+    CGPoint endCameraPoint = CGPointMake(centerPoint.x, centerPoint.y);
+    CLLocationCoordinate2D panCoordinate = [self convertPoint:endCameraPoint toCoordinateFromView:pan.view];
+    panCamera.centerCoordinate = panCoordinate;
+    
+    return panCamera;
+}
+
+- (MGLMapCamera *)currentCameraWithEstimatedZoom:(double)zoom
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: A pinch gesture performs a zoom anchored at the center of the pinch gesture. So this method needs something similar to the rotation method.

File: platform/ios/src/MGLMapView.mm (hunk 16 above)
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: If moving these methods to MGLMapCamera would be too much hassle because of dependencies on internal MGLMapView methods, we can leave the methods here, but let’s give them better names:

* `-currentCameraWithEstimatedEndPoint:panGesture:` → `-cameraByPanningWithTranslation:`
* `-currentCameraWithEstimatedZoom:` → `-cameraByZoomingToZoomLevel:`
* `-currentCameraWithEstimatedDegrees:anchorPoint:` → `-cameraByRotatingToDirection:aroundAnchorPoint:`
* `-currentCameraWithEstimatedPitch:anchorPoint:` → `-cameraByTiltingToPitch:`

File: platform/ios/src/MGLMapView.mm [hunk 19]
```diff
@@ -1619,6 +1736,86 @@ - (void)handleTwoFingerDragGesture:(UIPanGestureRecognizer *)twoFingerDrag
...
+    
+    mbgl::ScreenCoordinate anchor = mbgl::ScreenCoordinate { anchorPoint.x, anchorPoint.y };
+    currentCameraOptions.pitch = pitch * mbgl::util::DEG2RAD;
+    currentCameraOptions.anchor = anchor;
+    camera = [self cameraForCameraOptions:currentCameraOptions];
+    
+    return camera;
+}
+
+- (double)currentCameraZoom
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: This method is unnecessary; use `self.zoomLevel` instead.

File: platform/ios/src/MGLMapView.mm [hunk 20]
```diff
@@ -1619,6 +1736,86 @@ - (void)handleTwoFingerDragGesture:(UIPanGestureRecognizer *)twoFingerDrag
...
+    mbgl::CameraOptions currentCameraOptions = _mbglMap->getCameraOptions(padding);
+    MGLMapCamera *camera;
+    
+    currentCameraOptions.zoom = mbgl::util::clamp(zoom, self.minimumZoomLevel, self.maximumZoomLevel);
+    camera = [self cameraForCameraOptions:currentCameraOptions];
+    
+    return camera;
+}
+
+- (MGLMapCamera *)currentCameraWithEstimatedDegrees:(CGFloat)degrees anchorPoint:(CGPoint)anchorPoint
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: Use `CLLocationDirection` to represent the map’s rotation in degrees.

File: platform/ios/src/MGLMapView.mm [hunk 21]
```diff
@@ -1216,8 +1219,17 @@ - (void)handlePanGesture:(UIPanGestureRecognizer *)pan
...
         CGPoint delta = [pan translationInView:pan.view];
-        _mbglMap->moveBy({ delta.x, delta.y });
-        [pan setTranslation:CGPointZero inView:pan.view];
+
+        toCamera = [self currentCameraWithEstimatedEndPoint:delta panGesture:pan];
+        
+        if ([self.delegate respondsToSelector:@selector(mapView:shouldChangeFromCamera:toCamera:)]
+            && ![self.delegate mapView:self shouldChangeFromCamera:oldCamera toCamera:toCamera])
+        {
+            self.camera = oldCamera;
```
Comment (fabian-guerra) [inline platform/ios/src/MGLMapView.mm]: Correct, this only emphases we won't perform any gesture.

Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: OK, let’s remove this line and similar lines, then. Setting the camera also cancels any pending animations, which could be unexpected.

File: platform/ios/src/MGLMapView.mm [hunk 22]
```diff
@@ -1275,9 +1297,19 @@ - (void)handlePinchGesture:(UIPinchGestureRecognizer *)pinch
...
 
         if (log2(newScale) < _mbglMap->getMinZoom()) return;
-
-        _mbglMap->setScale(newScale, mbgl::ScreenCoordinate { centerPoint.x, centerPoint.y });
-
+        
+        // Calculates the final camera zoom, has no effect within current map camera.
+        MGLMapCamera *toCamera;
+        double zoom = log2(newScale);
+        toCamera = [self cameraByZoomingToZoomLevel:zoom aroundAnchorPoint:centerPoint];
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: Declare and set `toCamera` on the same line.

File: platform/ios/src/MGLMapView.mm [hunk 23]
```diff
@@ -1216,8 +1219,17 @@ - (void)handlePanGesture:(UIPanGestureRecognizer *)pan
     else if (pan.state == UIGestureRecognizerStateChanged)
     {
         CGPoint delta = [pan translationInView:pan.view];
-        _mbglMap->moveBy({ delta.x, delta.y });
-        [pan setTranslation:CGPointZero inView:pan.view];
+
+        toCamera = [self cameraByPanningWithTranslation:delta panGesture:pan];
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: Move the declaration of `toCamera` down to this line.

File: platform/ios/src/MGLMapView.mm [hunk 24]
```diff
@@ -1275,9 +1297,19 @@ - (void)handlePinchGesture:(UIPinchGestureRecognizer *)pinch
         CGFloat newScale = self.scale * pinch.scale;
 
         if (log2(newScale) < _mbglMap->getMinZoom()) return;
-
-        _mbglMap->setScale(newScale, mbgl::ScreenCoordinate { centerPoint.x, centerPoint.y });
-
+        
+        // Calculates the final camera zoom, has no effect within current map camera.
+        MGLMapCamera *toCamera;
+        double zoom = log2(newScale);
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: Nit: declare this a few lines further up to avoid saying `log2(newScale)` twice.

File: platform/ios/src/MGLMapView.mm [hunk 25]
```diff
@@ -1508,18 +1573,32 @@ - (void)handleDoubleTapGesture:(UITapGestureRecognizer *)doubleTap
...
         CGPoint gesturePoint = [self anchorPointForGesture:doubleTap];
-
-        mbgl::ScreenCoordinate center(gesturePoint.x, gesturePoint.y);
-        _mbglMap->scaleBy(2, center, MGLDurationInSecondsFromTimeInterval(MGLAnimationDuration));
-
-        __weak MGLMapView *weakSelf = self;
-
-        [self animateWithDelay:MGLAnimationDuration animations:^
+        
+        MGLMapCamera *toCamera = [self cameraByZoomingToZoomLevel:newZoom aroundAnchorPoint:gesturePoint];
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: Nit: s/newZoom/self.zoomLevel + 1/

File: platform/ios/src/MGLMapView.mm [hunk 26]
```diff
@@ -1216,8 +1218,16 @@ - (void)handlePanGesture:(UIPanGestureRecognizer *)pan
     else if (pan.state == UIGestureRecognizerStateChanged)
     {
         CGPoint delta = [pan translationInView:pan.view];
-        _mbglMap->moveBy({ delta.x, delta.y });
-        [pan setTranslation:CGPointZero inView:pan.view];
+
+        MGLMapCamera *toCamera = [self cameraByPanningWithTranslation:delta panGesture:pan];
+        
+        if (![self.delegate respondsToSelector:@selector(mapView:shouldChangeFromCamera:toCamera:)] ||
+            ([self.delegate respondsToSelector:@selector(mapView:shouldChangeFromCamera:toCamera:)]
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: This is redundant: `!A || (A & C)` is equivalent to `!A || C`.
